infrastructure/src/main/java/org/corfudb/infrastructure/ManagementAgent.java

```diff
@@ -133,37 +139,54 @@
      * Initialization task.
      * This task is blocked until the management server is bootstrapped and has a connected runtime.
      * Performs recovery if required.
-     * Bootstraps the primary sequencer if the server has been bootstrapped.
-     * Initiates the failure detection and healing detection tasks.
+     * Initiates the monitoring services which performs failure detection and healing detection tasks.
      */
     private void initializationTask() {
         log.info("Start initialization task");
 
+        // Wait for management server bootstrapped.
         try {
             while (!shutdown && serverContext.getManagementLayout() == null) {
-                log.warn("Management Server waiting to be bootstrapped");
+                log.warn("initializationTask: Management Server waiting to be bootstrapped");
                 Sleep.MILLISECONDS.sleepRecoverably(CHECK_BOOTSTRAP_INTERVAL.toMillis());
             }
+        } catch (InterruptedException e) {
+            // Due to shutdown, which interrupts this thread
+            log.debug("initializationTask: Initialization task interrupted without " +
+                    "management server bootstrapped");
+            return;
+        }
 
-            if (shutdown){
-                log.info("Shutdown signal. Interrupt initialization");
+        // Retry layout recovery if need until success.
+        long recoveryAttempts = 0;
+        while (!shutdown && !recovered) {
+            try {
+                Layout layout = serverContext.copyManagementLayout();
+                if (runRecoveryReconfiguration(layout, getCorfuRuntime())) {
+                    // If recovery succeeds, reconfiguration was successful.
+                    // Save the latest management layout.
+                    serverContext.saveManagementLayout(getCorfuRuntime().getLayoutView().getLayout());
+                    log.info("initializationTask: Recovery completed");
+                    recovered = true;
+                }
+
+                log.error("initializationTask: Recovery failed {} times. Retrying in {}s.",
+                        ++recoveryAttempts, RECOVERY_RETRY_INTERVAL);
+                Sleep.MILLISECONDS.sleepRecoverably(RECOVERY_RETRY_INTERVAL.toMillis());
```

Comment:
Out of curiosity -- why is sleepRecoverably used here instead of sleepUninterruptibly?

Reply:
Because we could shutdown the managementAgent while it's doing recovery, shutdown() will interrupt() this thread and we want to catch and log it rather than throwing an UnrecoverableCorfuInterruptedError.